pypesto/hierarchical/optimal_scaling_approach/optimal_scaling_problem.py

```diff
@@ -0,0 +1,530 @@
+    ix_matrices_from_arrays,
+)
+from .optimal_scaling_parameter import OptimalScalingParameter
+
+try:
+    import amici
+    import petab
+    from petab.C import OBSERVABLE_ID
+except ImportError:
+    pass
+
+
+class OptimalScalingProblem(InnerProblem):
+    """Class of the Optimal Scaling inner subproblem."""
+
+    def __init__(
+        self,
+        xs: List[OptimalScalingParameter],
+        data: List[np.ndarray],
+        method: str,
+    ):
+        """Construct the Optimal Scaling inner subproblem.
+
+        Parameters
+        ----------
+            xs:
+                List of `OptimalScalingParameter`s of the subproblem.
+            data:
+                The data of the problem.
+            method:
+                A string representing the method of the Optimal Scaling approach, either 'reduced' or 'standard'.
+        """
+        super().__init__(xs, data)
+        self.groups = {}
+        self.method = method
+
+        for group in self.get_groups_for_xs(InnerParameterType.OPTIMALSCALING):
+            self.groups[group] = {}
+            xs = self.get_xs_for_group(group)
+            self.groups[group]['num_categories'] = len(
+                {x.category for x in xs}
+            )
+            self.groups[group]['num_datapoints'] = np.sum(
+                [
+                    np.sum([np.sum(ixs) for ixs in x.ixs])
+                    for x in self.get_cat_ub_parameters_for_group(group)
+                ]
+            )
+
+            self.groups[group]['surrogate_data'] = np.zeros(
+                self.groups[group]['num_datapoints']
+            )
+
+            self.groups[group]['num_inner_params'] = (
+                self.groups[group]['num_datapoints']
+                + 2 * self.groups[group]['num_categories']
+            )
+
+            self.groups[group]['num_constr_full'] = (
+                2 * self.groups[group]['num_datapoints']
+                + 2 * self.groups[group]['num_categories']
+            )  # - 1
+
+            self.groups[group]['lb_indices'] = list(
+                range(
+                    self.groups[group]['num_datapoints'],
+                    self.groups[group]['num_datapoints']
+                    + self.groups[group]['num_categories'],
+                )
+            )
+
+            self.groups[group]['ub_indices'] = list(
+                range(
+                    self.groups[group]['num_datapoints']
+                    + self.groups[group]['num_categories'],
+                    self.groups[group]['num_inner_params'],
+                )
+            )
+
+            self.groups[group]['C'] = self.initialize_c(group)
+
+            self.groups[group]['W'] = self.initialize_w(group)
+
+            self.groups[group]['Wdot'] = self.initialize_w(group)
+
+    @staticmethod
+    def from_petab_amici(
+        petab_problem: petab.Problem,
+        amici_model: 'amici.Model',
+        edatas: List['amici.ExpData'],
+        method: str = None,
```

Comment:
```suggestion
        method: str = REDUCED,
```

Reply:
This was a laugh, interesting that this came first to mind to do